Introduce v3 onion support

mig5 opened the pull request:
This PR introduces v3 onion support. I'm deliberately basing this on the receiver-mode-gui branch because it has so many other breaking changes that it's not worth branching off of develop branch anymore.

This PR:

1) favours v3 onions if the Tor version is new enough

2) Requires the (testing) user to `pip install pysha3` if they don't have it already (I encountered an issue on Debian 9 whereby, if the version of Python is too low, the logic that uses hashlib doesn't quite work (`hashlib.sha3_256`) , but pysha3 does. However, there's no deb or rpm for it, so it introduces the need to pip install at least one package on those platforms (for the first time) if building/testing from source. Or, more generally, introduces that dependency. Not sure how that will fly for Debian/Ubuntu .debs....

3) Deliberately hides the 'persistent URL' and 'stealth' features in the settings dialog (since we know those don't work properly with v3 onions) 

4) even if somehow persistent or stealth mode is enabled (think: previously-saved .json file, or command line flags), they are silently ignored (resulting in a non-persistent onion with no stealth)

Otherwise it contains the logic from my previous prop224 branch. Note the onionkey.py, which *pre-generates* private keys for both v3 and v2 now. This is necessary to interact with Stem with v3 keys, but even more significantly, it paves the way for saving that v3 private key to json once persistence is possible, *and* it paves the way for a 'dead man's switch' feature (reverse shutdown-timer: a *start-up* timer that allows the user to communicate a *future* onionshare URL yet to be published), which I will send a PR for later down the track.


I'm assuming it's not quite ready yet.

the v3 onion works for both share and receiver mode.

Welcome testing.

Once the persistence feature is released in Tor stable (expected in December), we can remove the logic that ignores-or-hides persistence mode. Likewise for stealth (Client Auth)

@Baccount

mig5 commented:
Modified this PR so that rather than hiding the entire persistence/stealth settings if the Tor version is high enough, the user can optionally 'Use legacy v2 onion' mode via a checkbox in the Settings Dialog.

When this setting is checked on, persistence and stealth options are shown.

When persistence or stealth options are enabled, the legacy mode cannot be switched off (because those settings currently only work with v2 mode).

When persistence or stealth is disabled, the user can uncheck 'use legacy v2 mode' again, and go back to using v3 onions by default (on the assumption their Tor version is new enough for v3 in the first place).

mig5 commented:
P.S the 'use legacy v2 onion' mode checkbox doesn't really belong in 'Sharing' section of the Settings dialog, since it affects receiver mode too. So really #722 would need to be merged first, then there'll probably be conflicts here needing adjustment.

Baccount commented:
When I open the settings panel, the app crashes every time.
```
Traceback (most recent call last):
  File "/Users/jackimoore/Downloads/onionshare-prop224_receiver_mode/onionshare_gui/onionshare_gui.py", line 318, in open_settings
    d = SettingsDialog(self.common, self.onion, self.qtapp, self.config, self.local_only)
  File "/Users/jackimoore/Downloads/onionshare-prop224_receiver_mode/onionshare_gui/settings_dialog.py", line 71, in __init__
    self.use_legacy_v2_onions_checkbox.setText(strings._("gui_use_legacy_v2_onions_checkbox", True))
  File "/Users/jackimoore/Downloads/onionshare-prop224_receiver_mode/onionshare/strings.py", line 61, in translated
    return strings[k]
KeyError: 'gui_use_legacy_v2_onions_checkbox'
Abort trap: 6
```
Here is the traceback.
@mig5

mig5 commented:
D'oh, I was working across macihnes and I forgot to commit the share/locale/en.json.

Fixed..

asn-d6 commented:
Hello, will test in the next 5-6 days max.

Baccount commented:
@mig5 When navigating to the V3 onion using TorBrowser 7.5.6, the Onion loads and looks like this.
<img width="1112" alt="screen shot 2018-08-22 at 1 59 39 pm" src="https://user-images.githubusercontent.com/10781882/44490668-cb49c900-a613-11e8-862a-2eb363db3627.png">

Baccount commented:
When I have onionshare set too connect to TorBrowser and I close TorBrowser and Onionshare, then open onionshare, it crashes every time. Here is the traceback.
```
[Aug 22 2018 14:04:15] SettingsDialog.__init__
Traceback (most recent call last):
  File "onionshare_gui/onionshare_gui.py", line 318, in open_settings
    d = SettingsDialog(self.common, self.onion, self.qtapp, self.config, self.local_only)
  File "onionshare_gui/settings_dialog.py", line 73, in __init__
    if Version(self.onion.tor_version) < Version('0.3.2.9'):
AttributeError: 'Onion' object has no attribute 'tor_version'
Abort trap: 6
```

Baccount commented:
@mig5 https://github.com/micahflee/onionshare/pull/731#issuecomment-415179747 crash still occurs even with #701

mig5 commented:
@Baccount that looks like a valid bug when using Tor Browser as connection method and having no connection. I think that predates this bug e.g you could probably trigger that via the 'Update Checker' button too in latest version of OnionShare? Haven't tried, but I will look at whether we can workaround it later today. Thanks for finding

mig5 commented:
the Tor version parsing depends on an already-connected Tor. I don't think I can work around that to parse the Tor version any other way. But obviously we need the SettingsDialog to open if there is no Tor connection (to allow users to change the connection method).

Maybe we should just show the 'use legacy v2 onions' checkbox regardless of the Tor version (e.g even if the Tor version can *only* support legacy v2). Because either way, the backend/CLI will ignore any attempt to use persistence/stealth regardless of what the json settings say, as a safety measure.

I'll push such a 'fix' for now, and state explicitly in the 'use legacy v2 onion' label that this is true even if unchecked, if the Tor version is too old.

mig5 commented:
Pushed a fix that doesn't try and do any clever show/hide of features that are v2-only, based on Tor version.

Instead the labels show what is v2-only or not (for now).

If persistent mode or stealth are enabled, then the legacy v2 checkbox is forced on and read-only, since they are v2-only features.

Baccount commented:
Receive mode says "Don't prevent users from guessing onionshare address" does not seem appropriate for V3 onions, because the V3 design prevents the HSDIR from collecting Onion addresses , the crypto used is still unbroken ` (replaced SHA1/DH/RSA1024 with SHA3/ed25519/curve25519)` lastly, brute forcing a 56 characters long onion may be near impossible.
@mig5

mig5 commented:
I guess I agree, but the feature really means 'don't use a slug', and I think that should still be optional for the end user.. however, 'Don't use a slug' isn't useful language for non-technical users which is why @micahflee used the language currently present.

I think I'll let Micah decide if slugs should be forcibly-disabled if v3 is in use, but again, there's no way for me to 'know' if v3 can be used right now because checking the Tor version can crash the app as you've already discovered.

I also think it's already 'fixed' by making the slug optional in share mode too, a PR for which I made in #722, maybe even there the language could be changed to 'Don't add a random path to the URL' or something that means more than 'slug' to less tech-savvy users...

Anyway for now let's keep this PR purely about whether v3 works and whether forcing v2 with/without persistence/stealth mode also still works :) we have tickets about the slug stuff in #594 and #713

mig5 commented:
Thanks @asn-d6 - we encounter some difficulties in PyQt5 trying to auto-resize the parent widgets according to the size of the child widgets in some cases.

I initially experimented with elided text which expanded as the parent window was resized, but then encountered PyQt5 stylesheet issues with the text which I couldn't resolve. Followed @micahflee's out-of-band advice to just force a wrap on the string, but you're right, I didn't think about when the window is stretched..

Curious, did you stretch the window manually or did you somehow encounter it auto-expanding like that?

We can probably do a bit better at this anyway before merging, indeed..

asn-d6 commented:
> Curious, did you stretch the window manually or did you somehow encounter it auto-expanding like that?

I did not stretch the window manually. I use a tiling WM (i3) and it just started up auto-stretched. :face_with_head_bandage:

mig5 commented:
Thanks @asn-d6 ! I totally forgot about tiling WMs!

I've pushed a fix that only wraps it if the window size is below a certain size where QT's WordWrap can't sanely wrap it properly.

asn-d6 commented:
Works great, @mig5. Thanks!

micahflee commented:
Also about pysha3, it's packaged in Debian testing (buster):
https://packages.debian.org/buster/python3-sha3

And it's also in Ubuntu 18.04 (bionic), which is the new LTS so everyone should upgrade:
https://packages.ubuntu.com/bionic/python3-sha3

But also, I'm actually ok with not supporting Debian stable for OnionShare 2.0, just because software in stable is always very old, and it's easy enough to use testing if you're interested in using newer software. Also, we can avoid having any pip dependencies this way.

mig5 commented:
For Fedora I can't find an equivalent python3-sha3 package, at least not in Fedora 27, maybe later? Finding it harder to search than in debian repo pages.

micahflee commented:
Sorry, I just merged develop into receiver-mode-gui. That's why you have conflicts now. (Once #722 is merged, I think it's about time to merge receiver-mode-gui into develop, and then we can just start working off develop again...)

Yeah I also can't find pysha3 in Fedora 27 just by searching their package websites.

I tested this branch in Ubuntu 18.04, and it worked fine without having the python3-sha3 package instead, I guess because the version of python3 was new enough? I don't have a F27 VM handily available right now -- does it also have a new enough version of python3 so that we don't need the pysha3 dependency at all?

mig5 commented:
I'll install this branch quickly on a Fedora DispVM and test that.

And I'll fix these new conflicts here after #722 is merged (at which point I expect more conflicts anyway).

mig5 commented:
I confirm Fedora 27 doesn't need pysha3 installed (Python version is high enough).

The catch-22 is that Fedora 27's Tor binary is too old for v3 anyway, lol. I had to install Tor Browser and use that as my Tor connection method in order to even test v3 style onions on Fedora 27.

micahflee commented:
Haha, well ok then. I'm fine with basically requiring Debian testing, Ubuntu 18.04, or Fedora 28 for OnionShare 2.0 then. Do you want to remove the references to pysha3? We'll have to remember to update the [Linux Distribution Support](https://github.com/micahflee/onionshare/wiki/Linux-Distribution-Support) wiki page (and probably the whole wiki) for the 2.0 release.

Also, I've merged #722 now, so you can finish up this PR. Getting close!

mig5 commented:
Thanks @micahflee - argh, this is so close. It turns out that a QCheckbox can't be rich text (contain a hyperlink). Instead they still need to be QLabels in their own right.

I initially placed them underneath the checkbox but didn't like the look. So I've added them  as widgets inside a QHBoxLayout (and then add a widget with that layout), but now the checkboxes are indented compared to the other checkboxes. Do you want to copy my branch and fix if it's easy for you?

mig5 commented:
@micahflee NM, I fixed it. I think this is now ready (albeit the tests fail, but I think that's due to the public mode 404 route stuff)

micahflee commented:
Can you update `stdeb.cfg` to use these? We now require python 3.6+ because of the sha3 stuff. (And now that Ubuntu bionic is the new LTS the older versions are obsolete, I can't make a PPA release to older versions.)

```
Suite: bionic
X-Python3-Version: >= 3.6
```

micahflee commented:
Also, I found a bug. I just started OnionShare in share mode, public mode (also I started with `--local-only`). My OnionShare URL is http://127.0.0.1:17650/, but the download link in the corner links to http://127.0.0.1:17650/kitty-guts/download -- it includes the slug even though it shouldn't. And clicking that brings me to a 404 error.

Receive mode handles this by passing different paths into the template (for the action attr of the upload and close forms):

```python
if self.common.settings.get('public_mode'):
    upload_action = '/upload'
    close_action = '/close'
else:
    upload_action = '/{}/upload'.format(self.slug)
    close_action = '/{}/close'.format(self.slug)
```

We'll have to do something similar for the download link in the send template.

micahflee commented:
Also I like the https://github.com/micahflee/onionshare/wiki/Legacy-Addresses wiki page!

micahflee commented:
And finally a few more things about the settings dialog.

We're mixing and matching the term "settings" and "options" -- maybe we should just stick to "settings" since it's the "Settings" dialog. Can you change the headers to "General", "Sharing", and "Receiving"?

The auto-stop timer is actually a general setting, not just a sharing setting. Can you move it into the "General" section?

Can you rename "OnionShare is open to the public" to "Public mode (what's this?)" with a link to the wiki (I'll go through and thoroughly update the wiki before the 2.0 release too!) -- it just seems odd to say "OnionShare is" when really we're referring to the specific share. And even saying "don't prevent people from guessing the OnionShare address" is confusing without explaining slugs and rate-limiting.

Can you re-order the General settings to this?

- "Public mode (what's this?)"
- "Use auto-stop timer (what's this?)" -- we might as well link to the wiki for every setting, no?
- "Use legacy addresses (what's this?)"
- "Use a persistent address (legacy) (what's this?)"
- "Create stealth onion services (legacy) (what's this?)"

Finally, I think it's reasonable now to hide the legacy settings. When "Use legacy addreses" is unchecked, the two legacy settings can be totally hidden. When it's checked, they can be shown.

mig5 commented:
> Also, I found a bug. I just started OnionShare in share mode, public mode (also I started with --local-only). My OnionShare URL is http://127.0.0.1:17650/, but the download link in the corner links to http://127.0.0.1:17650/kitty-guts/download -- it includes the slug even though it shouldn't. And clicking that brings me to a 404 error.

I don't reproduce this :( it works as expected for me both local-only and real mode. The Download Files button is just a route to /download.

The template already has this:

```
                {% if slug %}
                <li><a class="button" href='/{{ slug }}/download'>Download Files</a></li>
                {% else %}
                <li><a class="button" href='/download'>Download Files</a></li>
                {% endif %}
```

mig5 commented:
@micahflee thanks for the thorough review. I think I've addressed all the issues. Note as above, I couldn't reproduce the slug appearing in the Download page when using public mode..

Can you find a way to align all those 'what's this' labels to the right? addStretch() doesn't seem to be doing anything here for some reason.

maqp commented on onionshare/onionkey.py:
```diff
@@ -0,0 +1,129 @@
+                                           backend=default_backend())
+    hs_public_key = private_key.public_key()
+
+    # Pre-generate the public key (.onion hostname)
```
Unlike with v3 services, the address of v2 services is not the public key, just a truncated hash of it. AFAIK the domain name is called "service ID" by the Tor devs.
